precompiles/erc20/approve_test.go

```diff
@@ -0,0 +1,562 @@
+		},
+		{
+			name: "fail - decrease allowance without existing authorization",
+			malleate: func() []interface{} {
+				return []interface{}{
+					s.keyring.GetAddr(1), big.NewInt(decreaseAmount),
+				}
+			},
+			errContains: "does not exist or is expired",
+		},
+		{
+			name: "pass - decrease allowance with existing authorization",
+			malleate: func() []interface{} {
+				s.setupSendAuthz(
+					s.keyring.GetAccAddr(1),
+					s.keyring.GetPrivKey(0),
+					sdk.NewCoins(sdk.NewInt64Coin(s.tokenDenom, amount)),
+				)
+
+				return []interface{}{
+					s.keyring.GetAddr(1), big.NewInt(decreaseAmount),
+				}
+			},
+			expPass: true,
+			postCheck: func() {
+				s.requireSendAuthz(
+					s.keyring.GetAccAddr(1),
+					s.keyring.GetAccAddr(0),
+					sdk.NewCoins(sdk.NewInt64Coin(s.tokenDenom, decreaseAmount)),
+					[]string{},
+				)
+			},
+		},
+		{
+			// TODO: do we want this to fail or should it delete the authorization? currently fails
+			name: "fail - decrease to zero and delete existing authorization",
+			malleate: func() []interface{} {
+				s.setupSendAuthz(
+					s.keyring.GetAccAddr(1),
+					s.keyring.GetPrivKey(0),
+					sdk.NewCoins(sdk.NewInt64Coin(s.tokenDenom, amount)),
+				)
+
+				return []interface{}{
+					s.keyring.GetAddr(1), big.NewInt(amount),
+				}
+			},
+			errContains: "spend limit must be positive",
+			postCheck: func() {
+				// NOTE: since the authorization is not deleted, we check that it still exists
+				s.requireSendAuthz(
+					s.keyring.GetAccAddr(1),
+					s.keyring.GetAccAddr(0),
+					sdk.NewCoins(sdk.NewInt64Coin(s.tokenDenom, amount)),
+					[]string{},
+				)
+			},
+		},
+		{
+			name: "fail - decrease allowance with existing authorization but decreased amount too high",
+			malleate: func() []interface{} {
+				s.setupSendAuthz(
+					s.keyring.GetAccAddr(1),
+					s.keyring.GetPrivKey(0),
+					sdk.NewCoins(sdk.NewInt64Coin(s.tokenDenom, amount)),
+				)
+
+				return []interface{}{
+					s.keyring.GetAddr(1), big.NewInt(amount + 1),
+				}
+			},
+			errContains: "subtracted value cannot be greater than existing allowance",
+		},
+		{
+			name: "fail - decrease allowance with existing authorization in different denomination",
+			malleate: func() []interface{} {
+				s.setupSendAuthz(
+					s.keyring.GetAccAddr(1),
+					s.keyring.GetPrivKey(0),
+					sdk.NewCoins(sdk.NewInt64Coin(s.bondDenom, amount)),
+				)
+
+				return []interface{}{
+					s.keyring.GetAddr(1), big.NewInt(decreaseAmount),
+				}
+			},
+			// TODO: have more verbose error message here like "authorization for different denomination found"?
+			errContains: "subtracted value cannot be greater than existing allowance",
+			postCheck: func() {
+				// NOTE: Here we check that the authorization for the other denom was not deleted
+				s.requireSendAuthz(
+					s.keyring.GetAccAddr(1),
+					s.keyring.GetAccAddr(0),
+					sdk.NewCoins(sdk.NewInt64Coin(s.bondDenom, amount)),
+					[]string{},
+				)
+			},
```

Comment:
do we want to have a clearer message here? Specifying that the existing authorization is for a different denomination instead of just giving the subtracted value error